railties/lib/rails/application/configuration.rb

```diff
@@ -277,6 +278,10 @@ def load_defaults(target_version)
             }
           end
 
+          if respond_to?(:active_job)
+            active_job.use_big_decimal_serializer = false
```

Comment:
I'm very confused with this. Why it is false here but true in the `new_freamework_defaults`? If it is a new framework default it should be true in `load_defaults(7.1)`. If it isn't a new framework default, it should not be set here, neither in the generated initializer.

Reply:
Yeah, it's my fault for not catching this in review.

Reply:
Ah, now I got it. Neverming.

Reply:
That is what I get by not paying attention to Rails for more than 3 months. I don't remember anything. Yeah, this config should be true with the `load_defaults(7.1)`. It should not be true with `load_defaults(7.0)`.

Reply:
Just to confirm we're on the same page:

- The next release of Rails will be 7.1
- It is unsafe to enable `BigDecimalSerializer` without an intermediate deployment first, so we need a config to control this
- Application owners should be encouraged to enable it **following a successful deploy** of 7.1 (or `main`)
- In the release **following 7.1** (7.2 or 8.0), we will remove the config and `BigDecimalSerializer` will always be enabled (_effectively making it `true`_)

Therefore,

- When using the 7.0 defaults, this should be disabled
- When using the 7.1 defaults, this should be enabled

Is this accurate, and if so, is the current implementation correct? If it isn't, what needs to change?

Reply:
This is accurate, but the current implementation isn't doing what you are describing. It is doing:

```
When using the 7.1 defaults, this config is disabled
```

What needs to change is this line from `false` to `true`

```ruby
active_job.use_big_decimal_serializer = true
```

Reply:
Okay, thanks @rafaelfranca! I think/hope I've wrapped my head around the way this is supposed to work.
1. The default value should be set to what we want new apps to use.
    ```ruby
    # activejob/lib/active_job.rb
    module ActiveJob
      singleton_class.attr_accessor :use_big_decimal_serializer
      self.use_big_decimal_serializer = true
    ```
2. The `load_defaults` method should allow existing application owners to request the 7.0 behavior.
    ```ruby
    # railties/lib/rails/application/configuration.rb
    module Rails
      class Application
        class Configuration < ::Rails::Engine::Configuration
          def load_defaults(target_version)
            case target_version.to_s
            # ...
            when "7.0"
              load_defaults "6.1"
              # ...
              if respond_to?(:active_job)
                active_job.use_big_decimal_serializer = false
              end
              # ...
    ```
    1. Requesting the 7.1 behavior should override the 7.0 behavior.
        ```ruby
        # railties/lib/rails/application/configuration.rb
                # ...
                when "7.1"
                  load_defaults "7.0"
                  # ...
                  if respond_to?(:active_job)
                    active_job.use_big_decimal_serializer = true
                  end
                  # ...
        ```
3. The `new_framework_defaults` for 7.1 should mention enabling the setting.
    ```ruby
    # railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults_7_1.rb.tt  
    # ...
    # Rails.application.config.active_job.use_big_decimal_serializer = true
    ```
4. In Rails 7.2/8.0 if we choose to remove the config entirely, we would delete all mentions of the config from these files and keep only the `true` behavior in the implementation.

Have we documented this anywhere for Rails contributors? If not, it may be useful to do so, as it was not obvious how this was supposed to work. I'm glad to open a PR for that documentation if pointed in to the right location.

Reply:
> The default value should be set to what we want new apps to use

No, the default value should be set to what existing apps expect (the old, compatible, value).

> The load_defaults method should allow existing application owners to request the 7.0 behavior

Given the above, no-one needs to request the 7.0 behaviour because it's the default.

> Requesting the 7.1 behavior should override the 7.0 behavior.

Yes; that's what ensures newly-generated apps get the new behaviour.

> The new_framework_defaults for 7.1 should mention enabling the setting.

Yes; that's what prompts upgrading apps to consider switching to the new behaviour.

> if we choose to remove the config entirely, we would delete all mentions of the config from these files and keep only the true behavior in the implementation.

To remove the config option in the future, we'd deprecate enqueuing a BigDecimal with it set to `false` for a version -- after that we can assume that everyone has it set to true (whether via `load_defaults 7.1` or an explicit setting), and remove the false behaviour.

I can't think of anywhere we've explicitly documented this process... perhaps an internal comment on `load_defaults` would work? It feels too detailed and niche to drop into the Contributing guide.

Reply:
> .. perhaps an internal comment on `load_defaults` would work?

I think so. It's a thing we very often have to explain, and there's several moving pieces (new_framework_defaults, configuring.md, etc), it's frequent we forget one.

Reply:
Okay, I've pushed commits to #45664 which fix this, as well as add a documentation comment within the method. We can further discuss there, if needed.

Thank you all for your detailed explanations! 🙇